src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,350 @@
+                Debug.Assert(charCount < unprocessedBufferToBeRead.Length);
+
+                CurrentEncoding.GetChars(buffer, 0, buffer.Length, unprocessedBufferToBeRead, ++startIndex);
+                endIndex += charCount;
+
+            }
+        }
+
+        public override string ReadLine()
+        {
+            StringBuilder sb = new StringBuilder();
+
+            //Check if there is anything left in the unprocessedBufferToBeRead.
+            while (!IsExtraBufferEmpty())
+            {
+                ConsoleKeyInfo keyInfo = ReadKeyInternal();
+
+                if (keyInfo.Key == ConsoleKey.Enter)
+                    return sb.ToString();
+
+                else if (keyInfo.Key == ConsoleKey.Backspace && sb.Length > 0)
+                    sb.Length--;
+
+                if (keyInfo.KeyChar != '\0') // We ignore all non-printable keys
+                    sb.Append(keyInfo.KeyChar);
+            }
+
+            return sb.Append(base.ReadLine()).ToString();
+        }
+
+
+        public override int Read()
+        {
+            int x;
+
+            // Convert byte to char.
+            if (IsExtraBufferEmpty())
+            {
+                x = base.Read();
+            }
+            else
+            {
+                x = (int)unprocessedBufferToBeRead[startIndex++];
+            }
+
+            return x;
+        }
+
+        internal ConsoleKey GetKeyFromCharValue(ref char x, out bool isShift, out bool isCtrl)
+        {
+            isShift = false;
+            isCtrl = false;
+
+            switch (x)
+            {
+                case '\b':
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Backspace;
+
+                case '\t':
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Tab;
+
+                case (char)(int)(0x0A):
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Enter;
+
+                case (char)(int)(0x1B):
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Escape;
+
+                case (char)(int)(0x2A): return ConsoleKey.Multiply;
+
+                case (char)(int)(0x2B): return ConsoleKey.Add;
+
+                case (char)(int)(0x2D): return ConsoleKey.Subtract;
+
+                case (char)(int)(0x2F): return ConsoleKey.Divide;
+
+                case (char)(int)(0x2E): return ConsoleKey.Decimal;
+
+                case (char)(int)(0x7F):
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Delete;
+
+                case (char)(int)(0x20): return ConsoleKey.Spacebar;
+
+                case (char)(int)(0x21):
+                    isShift = true;
+                    return ConsoleKey.D1;
+                    
+                case (char)(int)(0x40):
+                    isShift = true;
+                    return ConsoleKey.D2;
```

Comment:
@pallavit I actually can't think of a way to do it as well, I just wanted to highlight it and make sure we're fine with this limitation :)

Btw. Mono seems to never return `Decimal` (it does `Multiply`, `Add`, etc.) so that may be another option.